sycl/include/sycl/types.hpp

```diff
@@ -905,11 +905,42 @@ template <typename Type, int NumElements> class vec {
   // Implement operator [] in the same way for host and device.
   // TODO: change host side implementation when underlying type for host side
   // will be changed to std::array.
-  const DataT &operator[](int i) const {
+  // NOTE: aliasing on bfloat16 may lead to problems if aggressively optimized.
+  // specializing with noinline to avoid as workaround.
```

Comment:
Could you please elaborate on the problems?

Reply:
According to the information @cperkinsintel shared offline, current implementation of `operator[]` introduces undefined behavior because the same data is accessed using two "incompatible" types (https://stackoverflow.com/a/37119041).

I don't think restricting inlining is a good workaround for that type of problems due to potential performance hit. Can we memcpy DataT local variable instead?

Reply:
I tried memcpy but the problem here is that operator[] returns a reference, not a value. And we specialize the overload of vec<bfloat16> specifically to limit the performance exposure.

Reply:
I see. Thank you for clarifying.